core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,406 @@
+            writerPhysicalWrittenBytesAtLastRebalance.set(i, writerPhysicalWrittenBytes.get(i));
+        }
+    }
+
+    private class RebalanceContext
+    {
+        private final Set<Integer> rebalancedPartitions = new HashSet<>();
+        private final long[] writerPhysicalWrittenBytesSinceLastRebalance;
+        private final long[] writerRowCountSinceLastRebalance;
+        private final long[] writerPriorities;
+        private final List<IndexedPriorityQueue<PartitionIdWithRowCount>> writerMaxPartitions;
+
+        private RebalanceContext(List<Long> writerPhysicalWrittenBytes, Map<WriterPartitionId, Long> partitionsRowCount)
+        {
+            writerPhysicalWrittenBytesSinceLastRebalance = new long[numberOfWriters];
+            writerPriorities = new long[numberOfWriters];
+            for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+                long physicalWrittenBytesSinceLastRebalance =
+                        writerPhysicalWrittenBytes.get(writerId) - writerPhysicalWrittenBytesAtLastRebalance.get(writerId);
+                writerPhysicalWrittenBytesSinceLastRebalance[writerId] = physicalWrittenBytesSinceLastRebalance;
+                writerPriorities[writerId] = physicalWrittenBytesSinceLastRebalance;
+            }
+
+            writerRowCountSinceLastRebalance = new long[numberOfWriters];
+            partitionsRowCount.forEach((writerPartitionId, rowCount) -> writerRowCountSinceLastRebalance[writerPartitionId.writerId] += rowCount);
+
+            writerMaxPartitions = new ArrayList<>(numberOfWriters);
+            for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+                writerMaxPartitions.add(new IndexedPriorityQueue<>());
+            }
+
+            partitionsRowCount.forEach((writerPartitionId, rowCount) ->
+                    writerMaxPartitions
+                            .get(writerPartitionId.writerId)
+                            .addOrUpdate(new PartitionIdWithRowCount(writerPartitionId.partitionId, rowCount), rowCount));
+        }
+
+        private boolean rebalancePartition(WriterId from, WriterId to)
+        {
+            IndexedPriorityQueue<PartitionIdWithRowCount> maxPartitions = writerMaxPartitions.get(from.id);
+
+            while (true) {
+                // Find the partition with maximum written bytes since last rebalance
+                PartitionIdWithRowCount partitionToRebalance = maxPartitions.poll();
+
+                if (partitionToRebalance == null) {
+                    return false;
+                }
+
+                PartitionInfo partitionInfo = partitionInfos[partitionToRebalance.id];
+
+                // If partition is already balanced or min skewed writer is already writing to that partition, then skip
+                // this partition and move on to the other partition inside max writer.
+                if (rebalancedPartitions.contains(partitionToRebalance.id) || partitionInfo.containsWriter(to.id)) {
```

Comment:
nit: We could use `isPartitionRebalanced` instead of `rebalancedPartitions.contains(partitionToRebalance.id)`